File: June-LeetCoding-Challenge-2020/Week5/62-Unique-Paths.cpp
// TC: O(m*n)
// SC: O(m*n)
// Runtime: 8 ms, faster than 14.01%
// Memory Usage: 6.1 MB, less than 66.14%
class Solution {
public:
    int uniquePaths(int m, int n) {
        
        int array[m][n];
        
        array[0][0] = 1;
        
        for(int i = 1; i < m; i++)
            array[i][0] = 1;
        
        for(int i = 1; i < n; i++)
            array[0][i] = 1;
        
        for(int i = 1; i < m; i++)
            for(int j = 1; j < n; j++){
                array[i][j] = array[i-1][j] + array[i][j-1];
            }
        
        return array[m-1][n-1];
                
    }
};

File: June-LeetCoding-Challenge-2020/Week5/62-unique-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @return {number}
 * Runtime: 68 ms, faster than 57.22% of JavaScript online submissions for Unique Paths.
 * Memory Usage: 34 MB, less than 55.77% of JavaScript online submissions for Unique Paths.
 * TC: O(m * n)
 * SC: O(m * n)
 */
var uniquePaths = function(m, n) {
    const dp = [];

    for (let i = 0; i < m; i++) {
        dp[i] = [1];
    }

    for (let i = 1; i < n; i++) {
        dp[0][i] = 1;
    }

    for (let col = 1; col < m; col++) {
        for (let row = 1; row < n; row++) {
            dp[col][row] = dp[col - 1][row] + dp[col][row - 1];
        }
    }

    return dp[m - 1][n - 1];
};

